soliket/cosmopower.py

```diff
@@ -69,38 +90,82 @@ def calculate(self, state, want_derived=True, **params):
                         cmb_params[par] = [params[r]]
                         break
 
-        state["tt"] = self.cp_tt_nn.ten_to_predictions_np(cmb_params)[0, :]
-        state["te"] = self.cp_te_nn.predictions_np(cmb_params)[0, :]
-        state["ee"] = self.cp_ee_nn.ten_to_predictions_np(cmb_params)[0, :]
-        state["ell"] = self.cp_tt_nn.modes
+        ells = None
+
+        for spectype in self.networks:
+            network = self.networks[spectype]
+            used_params = { par : (cmb_params[par] if par in cmb_params else [params[par]]) for par in network["parameters"] }
+
+            if network["log"]:
+                data = network["network"].ten_to_predictions_np(used_params)[0, :]
+            else:
+                data = network["network"].predictions_np(used_params)[0, :]
+
+            state[spectype] = data
+
+            if ells is None:
+                ells = network["network"].modes
+
+        state["ell"] = ells.astype(int)
 
-    def get_Cl(self, ell_factor=False, units="FIRASmuK2"):
+    def get_Cl(self, ell_factor:bool =False, units:str ="FIRASmuK2") -> dict:
         cls_old = self.current_state.copy()
 
-        lmax = self.extra_args["lmax"] or (cls_old["tt"].shape[0] + 2)
+        lmax = self.extra_args["lmax"] or cls_old["ell"].max()
 
-        cls = {"ell": np.arange(lmax).astype(int)}
+        cls = {"ell": np.arange(lmax+1).astype(int)}
         ls = cls_old["ell"]
+        ls_fac = np.ones_like(ls).astype(float)
 
-        for k in ["tt", "te", "ee"]:
-            # cls[k] = np.zeros(cls["ell"].shape, dtype=float)
-            cls[k] = np.empty(cls["ell"].shape, dtype=float)
-            cls[k][:] = np.nan
+        for k in self.networks:
+            cls[k] = np.tile(np.nan, cls["ell"].shape)
 
-        cmb_fac = self._cmb_unit_factor(units, 2.7255)
+        for k in self.networks:
+            prefac = np.ones_like(ls).astype(float)
 
-        if ell_factor:
-            ls_fac = ls * (ls + 1.0) / (2.0 * np.pi)
-        else:
-            ls_fac = 1.0
+            if self.networks[k]["has_ell_factor"]:
+                prefac /= self.ell_factor(ls, k)
+            if ell_factor:
+                prefac *= self.ell_factor(ls, k)
 
-        for k in ["tt", "te", "ee"]:
-            cls[k][ls] = cls_old[k] * ls_fac * cmb_fac ** 2.0
+            cls[k][ls] = cls_old[k] * prefac * self.cmb_unit_factor(k, units, 2.7255)
+            cls[k][:2] = 0.0
             if np.any(np.isnan(cls[k])):
                 self.log.warning("CosmoPower used outside of trained "\
                                  "{} ell range. Filled in with NaNs.".format(k))
 
         return cls
 
-    def get_can_support_params(self):
-        return ["omega_b", "omega_cdm", "h", "logA", "ns", "tau_reio"]
+    def ell_factor(self, ls: np.ndarray, spectra: str) -> np.ndarray:
```

Comment:
Maybe mention somewhere (either here or in the test dict) that these ell_factor conversions specifically refer to CLASS units.

Reply:
They refer to the generic Cl/Dl prefactor, see for example [here](https://camb.readthedocs.io/en/latest/results.html#camb.results.CAMBdata.get_cmb_power_spectra). They are used to resolve the `ell_factor` option in `get_Cl` which is [defined in BoltzmannBase](https://github.com/CobayaSampler/cobaya/blob/master/cobaya/theories/cosmo/boltzmannbase.py#L312). (I have never used class so I don't refer to anything from class specifically).